src/full/Agda/Utils/IORef/Strict.hs

```diff
@@ -0,0 +1,222 @@
+--   avoid doing redundant work, though warnings about repeated read/writes and memory
+--   leaks still apply.
+--
+-- == Caveats
+--
+-- The API for 'StrictIORef' only ensures that the value in the reference is in
+-- weak-head normal form, *not* normal form. This is done for the sake of efficiency:
+-- using 'Control.DeepSeq.NFData' to place the value into normal form requires us to
+-- fully traverse the value being stored inside of the reference, which can be quite
+-- expensive for recursive structures.
+--
+-- This design decision does lead to some unfortunate gotchas. Namely, if you store
+-- a type inside of a 'StrictIORef' that has non-strict fields, it is still possible
+-- to have space/memory leaks. This is best explained by an example:
+--
+-- @
+-- gotcha :: IO ()
+-- gotcha = do
+--   ref <- newStrictIORef (Just 1)
+--   loop ref (10 ^ 9)
+--   n <- readStrictIORef ref
+--   print n
+--   where
+--     loop :: StrictIORef (Maybe Int) -> Int -> IO ()
+--     loop ref i =
+--       if i == 0 then
+--         pure ()
+--       else do
+--         n <- readStrictIORef ref
+--         writeStrictIORef ref (fmap (+ 1) n)
+--         loop ref (i - 1)
+-- @
+--
+-- This leaks even though we are using 'StrictIORef', as 'writeStrictIORef' only forces the
+-- thunk containing the 'Just', and *not* the thunk stored within the 'Just'. Similar problems
+-- exist with 'StrictIORef (a, b)', as pairs in Haskell are lazy.
+--
+-- In light of this, users are encouraged to use 'StrictIORef' in concert with types that have strict
+-- fields. This ensures that forcing the outer constructor of the type also forces the fields, avoiding any potential
+-- leaks. If this is not possible, users are encouraged to use strict let bindings, 'seq', and
+-- extreme care.
+--
+-- == Implementation notes
+--
+-- When forcing values to WHNF before writing to the reference, we have two options
+--
+-- * Use @writeIORef ref $! a@
+-- * Use @writeIORef ref =<< evaluate a@
+--
+-- These are subtly different in the presence of exceptions. The former
+-- unfolds to:
+--
+-- @
+-- let !a' = a
+-- in writeIORef ref a
+-- @
+--
+-- If forcing @a@ to WHNF causes an exception to get thrown, then we don't produce any IO action at
+-- all. Conversely, @writeIORef ref =<< evaluate a@ will always produce an IO action that itself will
+-- throw when evaluated.
+--
+-- In practice, this means that
+-- @
+-- (writeIORef ref $! (error "a")) >> error "b"
+-- @
+-- can throw either @"a"@ or @"b"@ depending on what the optimizer decides to do, whereas
+-- @
+-- evaluate (error "a") >>= writeIORef ref >> error "b"
+-- @
+-- must always throw @"a"@.
+--
+-- We opt to use @writeIORef ref $! a@ in this module, as it gives the optimizer more leeway
+-- when doing code transformations that might move where a value gets forced (EG: worker-wrapper).
+-- This is a safe choice, as Agda does not rely on precise exception semantics for correctness.
```

Comment:
N.B.: Indeed not, only for debug printing, see e.g.: 
- https://github.com/agda/agda/issues/6728

(No need to change your formulation...)